core/Enclave/key_factory.cpp

```diff
@@ -0,0 +1,396 @@
+        return UINT32_MAX;
+
+    if (aad_size > UINT32_MAX - plaintext_size)
+        return UINT32_MAX;
+
+    if (sizeof(sgx_aes_gcm_data_ex_t) > UINT32_MAX - plaintext_size - aad_size)
+        return UINT32_MAX;
+
+    return (aad_size + plaintext_size + sizeof(sgx_aes_gcm_data_ex_t));
+}
+
+uint32_t ehsm_get_gcm_ciphertext_size(const sgx_aes_gcm_data_ex_t *gcm_data)
+{
+    if (NULL == gcm_data)
+        return UINT32_MAX;
+
+    return gcm_data->ciphertext_size;
+}
+
+// use the g_domain_key to encrypt the cmk and get it ciphertext
+sgx_status_t ehsm_create_keyblob(const uint8_t *plaintext, const uint32_t plaintext_size, 
+                                 const uint8_t *aad, const uint32_t aad_size, 
+                                 sgx_aes_gcm_data_ex_t *keyblob_data)
+{
+    uint32_t real_aad_size = aad_size;
+    if (NULL == aad)
+        real_aad_size = 0;
+
+    sgx_status_t ret = sgx_read_rand(keyblob_data->iv, sizeof(keyblob_data->iv));
+    if (ret != SGX_SUCCESS) {
+        printf("error generating iv.\n");
+        return ret;
+    }
+
+    ret = sgx_rijndael128GCM_encrypt(&g_domain_key,
+                                     plaintext, plaintext_size,
+                                     keyblob_data->payload,
+                                     keyblob_data->iv, sizeof(keyblob_data->iv),
+                                     &(keyblob_data->payload[keyblob_data->ciphertext_size]), real_aad_size,
+                                     reinterpret_cast<uint8_t (*)[16]>(keyblob_data->mac));
+    if (SGX_SUCCESS != ret) {
+        printf("gcm encrypting failed.\n");
+    }
+    else {
+        keyblob_data->ciphertext_size = plaintext_size;
+        keyblob_data->aad_size = real_aad_size;
+    }
+
+    return ret;
+}
+
+// use the g_domain_key to decrypt the cmk and get it plaintext
+sgx_status_t ehsm_parse_keyblob(uint8_t *plaintext, uint32_t plaintext_size,
+                                        const sgx_aes_gcm_data_ex_t *keyblob_data)
+{
+    if (NULL == keyblob_data || NULL == plaintext || NULL == plaintext_size
+                 || plaintext_size < keyblob_data->ciphertext_size)
+        return SGX_ERROR_INVALID_PARAMETER;
+
+    sgx_status_t ret = sgx_rijndael128GCM_decrypt(&g_domain_key,
+                                                  keyblob_data->payload, keyblob_data->ciphertext_size,
+                                                  plaintext,
+                                                  keyblob_data->iv, sizeof(keyblob_data->iv),
+                                                  &(keyblob_data->payload[keyblob_data->ciphertext_size]), keyblob_data->aad_size,
+                                                  (const sgx_aes_gcm_128bit_tag_t*)keyblob_data->mac);
+    if (SGX_SUCCESS != ret)
+        printf("gcm decrypting failed.\n");
+    else
+        plaintext_size = keyblob_data->ciphertext_size;
+
+    return ret;
+}
+
+uint32_t ehsm_get_symmetric_key_size(ehsm_keyspec_t key_spec)
+{
```

Comment:
seems this api is  useless

Reply:
This api will be used later

Reply:
it only used internally, why not define it as static and if so, then remove ehsm_xxx

Reply:
copy, it will change